pkg/modules/notary/notary.go

```diff
@@ -0,0 +1,300 @@
+		// with the associated fallback transactions grouped by the main transaction hash
+		requests map[util.Uint256]request
+
+		// accMtx protects account.
+		accMtx      sync.RWMutex
+		currAccount *wallet.Account
+		wallet      *wallet.Wallet
+	}
+
+	Config struct {
+		MainCfg config.P2PNotary
+		Chain   blockchainer.Blockchainer
+		Log     *zap.Logger
+	}
+)
+
+// request represents Notary service request.
+type request struct {
+	typ RequestType
+	// isValid indicates whether main transaction satisfies Notary node verification rules and can be sent to the network.
+	isValid bool
+	// isSent indicates whether main transaction was successfully sent to the network.
+	isSent bool
+	main   *transaction.Transaction
+	// minNotValidBefore is the minimum NVB value among fallbacks transactions.
+	// We stop trying to send mainTx to the network if the chain reaches minNotValidBefore height.
+	minNotValidBefore uint32
+	fallbacks         []*transaction.Transaction
+	// nKeys is the number of witnesses to be collected for the main transaction
+	nKeys uint8
+	// nKeysCollected is the number of already collected witnesses
+	nKeysCollected uint8
+}
+
+// NewNotary returns new Notary module.
+func NewNotary(bc blockchainer.Blockchainer, log *zap.Logger) (*Notary, error) {
+	cfg := bc.GetConfig().Modules.P2PNotary
+	w := cfg.UnlockWallet
+	wallet, err := wallet.NewWalletFromFile(w.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	haveAccount := false
+	for _, acc := range wallet.Accounts {
+		if err := acc.Decrypt(w.Password); err == nil {
+			haveAccount = true
+			break
+		}
+	}
+	if !haveAccount {
+		return nil, errors.New("no wallet account could be unlocked")
+	}
+
+	return &Notary{
+		requests: make(map[util.Uint256]request),
+		Config: Config{
+			MainCfg: cfg,
+			Chain:   bc,
+			Log:     log,
+		},
+		wallet: wallet,
+	}, nil
+}
+
+// OnNewRequest is a callback method which is called after new notary request is added to the notary request pool.
+func (n *Notary) OnNewRequest(payload *payload.P2PNotaryRequest) {
+	if n.getAccount() == nil {
+		return
+	}
+
+	n.reqMtx.Lock()
+	defer n.reqMtx.Unlock()
+	var isValid bool
+	nvbFallback := payload.FallbackTransaction.GetAttributes(transaction.NotValidBeforeT)[0].Value.(*transaction.NotValidBefore).Height
+	nKeys := payload.MainTransaction.GetAttributes(transaction.NotaryAssistedT)[0].Value.(*transaction.NotaryAssisted).NKeys
+	typ, err := n.verifyIncompleteWitnesses(payload.MainTransaction, nKeys)
+	if err == nil { // valid main tx signatures
+		isValid = true
+	}
+	r, ok := n.requests[payload.MainTransaction.Hash()]
+	if ok {
+		if r.nKeysCollected >= nKeys { // already collected sufficient number of keys to complete main transaction
+			return // TODO: however, we may still accept extra fallback transactions without adding new signatures to main tx
+		}
+		r.isValid = isValid && r.isValid && typ == r.typ // request for the main tx is found, but witnesses types may be different
```

Comment:
How can `typ != r.typ` happen? The main transaction is the same.

Reply:
Hashed part of the main transaction is the same. And witnesses are not included in the hashed part. So there might be a case when `main1` has standard witnesses to be collected, and `main2` has multisignature witness to be completed. In this case, we can't finalize the request correctly, so fallback transactions will be applied.

Reply:
`main1` and `main2` contain addresses which effectively are hashes of verification scripts, if you're in this situation one of transactions is wrong because its verification script doesn't match sender hash (so it should've been rejected long before getting here). Which returns us to the question of how can `typ != r.typ` happen?

Reply:
Got it, fixed.